### PR TITLE
Added unit tests for display.toggle_fullscreen and display.get_active

### PR DESCRIPTION
**Pull request by tsadama**
Hello, this pull request addresses #1743 and #1755. 

The unit tests created for display.toggle_fullscreen() and display.get_active() were created with stipulations from the documentation and from my understanding of the code. I used pylint to check my code.

Two things I noticed when testing locally:

- The documentation for display.get_active() states that it will return false when an image is iconified. However, iconify returned true, but get_active returned true still. 

- When running display.set_mode() with width/height of (0,0) and then display.toggle_fullscreen(), the resulting width/height becomes (800, 600) with a display.toggle_fullscreen() return value that is not 1. However for all other width/heights I tried, the resulting width/height is the same as the original with a display.toggle_fullscreen() return value that is not 1.

 Please let me know if I should make any changes!

### CODE REVIEW

**Comment by tsadama** [comment]
Follow up: it looks like one of the reasons for the unsuccessful build is because of the bullet point #1. Guidance on how I should address this would be appreciated!

**Comment by MyreMylar** [comment]
> Hello, this pull request addresses #1743 and #1755.
> 
> The unit tests created for display.toggle_fullscreen() and display.get_active() were created with stipulations from the documentation and from my understanding of the code. I used pylint to check my code.
> 
> Two things I noticed when testing locally:
> 
> * The documentation for display.get_active() states that it will return false when an image is iconified. However, iconify returned true, but get_active returned true still.

Since this is happening locally and nor just on CI it sounds like a bug with get_active(). One of the issues we are trying to solve with getting unit tests for everything is finding actual bugs lurking in the code. I will have a poke at this tomorrow.

Sometimes the Continuous Integration builds behave differently to local builds.

> * When running display.set_mode() with width/height of (0,0) and then display.toggle_fullscreen(), the resulting width/height becomes (800, 600) with a display.toggle_fullscreen() return value that is not 1. However for all other width/heights I tried, the resulting width/height is the same as the original with a display.toggle_fullscreen() return value that is not 1.

Interesting, I guess it can't really make a full screen display with 0 pixels to play with so picking a reasonable default seems OK to me?

**Comment by tsadama** [comment]
Thank you for the response. In the meantime, I will fix some of the other errors that popped up with the CI.

**Comment by MyreMylar** [comment]
I had a quick look at `get_active()` and remembered this old bug:

https://github.com/pygame/pygame/issues/161

Which, reading between the lines, seems to indicate that `get_active()` maybe doesn't work only on windows?

 Are you on windows @tsadama ?

**Comment by MyreMylar** [comment]
I'm using this MOR from the old bug:

```
import pygame


pygame.init()
window_surface = pygame.display.set_mode((640, 480))

is_running = True
while is_running:
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            is_running = False

    if not pygame.display.get_active():
        print(str(pygame.display.get_active()))

    pygame.display.update()

```

Last time @illume reported that it worked on mac and @nthykier reported that it worked on his machine, which I'm guessing is probably linux of some flavour?

**Comment by MyreMylar** [comment]
Tried that same code on Raspberry Pi OS (which is basically Debian) and it works fine. So it's looking like it is probably a windows issue.

**Comment by MyreMylar** [comment]
Poked at the C code some, it looks like on windows `SDL_WINDOW_SHOWN` is always true whatever the minimised state of the window. However, I notice that there is also another flag called `SDL_WINDOW_MINIMIZED` which does work on windows.

Perhaps you could change `display.c` to read something like:

    #if IS_SDLv2
    static PyObject *
    pg_get_active(PyObject *self, PyObject *args)
    {
        Uint32 flags = SDL_GetWindowFlags(pg_GetDefaultWindow());
        return PyBool_FromLong((flags & SDL_WINDOW_SHOWN) &&
                               !(flags & SDL_WINDOW_MINIMIZED));
    }
    #else  /* IS_SDLv1 */
    static PyObject *
    pg_get_active(PyObject *self, PyObject *args)
    {
        return PyBool_FromLong((SDL_GetAppState() & SDL_APPACTIVE) != 0);
    }
    #endif /* IS_SDLv1 */

And see if it makes your tests work locally at least?

**Comment by tsadama** [comment]
Hi @MyreMylar , I am actually using Ubuntu 16.04. I tried the MOR a few posts above, and my command line is being flooded with False so it seems like it's working. 

In terms of the change to the C code, I applied it and then did the following:
python3 setup.py build 
sudo python3 setup.py install 
ran the unit test on get_active

I still got the output where get_active would return true after a call to iconify, although I'm not entirely sure that I am rebuilding/recompiling it properly after changing the C code.

**Comment by MyreMylar** [comment]
Not looked over what you've done yet, but I wonder if a couple of calls to pygame.event.get() or pygame.event.pump() after calling iconify() will sort out the issue for the unit tests. I suspect that it might be the sort of SDL instruction where it sends a request to the OS and then waits for an event to come back to know if it worked.

**Comment by tsadama** [comment]
Follow up: 
I've tried modifying the C code as mentioned above and I've also tried updating the display, and using calls to pygame.event.get() and pygame.event.pump(). Unfortunately, I am unable to determine why iconify does not cause get_active() to be false. Could it be possible that iconify is buggy here? If so, it might be worth saving the iconify/get_active() check for an iconify unit test.

**Comment by MyreMylar** [comment]
> Could it be possible that iconify is buggy here? If so, it might be worth saving the >iconify/get_active() check for an iconify unit test.

I guess an iconify() test case would be useful to check that. Modifying the code above:

```
import pygame


pygame.init()
window_surface = pygame.display.set_mode((640, 480))

is_running = True
while is_running:
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            is_running = False
        if event.type == pygame.KEYDOWN:
            pygame.display.iconify()

    if not pygame.display.get_active():
        print(str(pygame.display.get_active()))

    pygame.display.update()
```

For me, on windows, that makes pressing any key minimize the window just as if I'd pressed the minimize button on the menu bar. It doesn't print False when minimized on dev10, but it does if I add the C code change mentioned above.

If .iconify() doesn't minimize the window on linux then I guess we have our answer.

**Comment by MyreMylar** [comment]
Just tried the above code using `pygame.display.iconify()` and it works fine on Raspberry Pi OS (which is Debian based) and using dev10. It prints False continuously after pressing a key to minimize the window.

**Comment by tsadama** [comment]
I tried the above code and it looks like it’s working! To test the code without manually pressing a key, I used pygame.event to post a keydown event. I will make a commit once I factor this in to the unit test and make sure everything’s working locally

**Comment by tsadama** [comment]
Update: the previous commits were my attempts to fix get_active. However, the CI builds are all terminating bc of a lack of output for 10 + minutes. I thought this might have been because of an infinite loop, but when I commented out the loop, the CI builds are still terminating.

**Comment by robertpfeiffer** [comment]
I would expect his test to fail if the user is using a tiling window manager.

**Comment by MyreMylar** [comment]
OK, I investigated this a bit. 

1. From what I can tell you can trim the failing part of the test right down to something like this:

```
pygame.init()

pygame.display.set_mode((640, 480))

pygame.event.clear()
pygame.display.iconify()

self.assertEqual(pygame.display.get_active(), False)
```

On windows and so long as you have also added the C code change above it will pass locally.

2. This part of the test at least will not work on the CI (Travis & Appveyor) as `iconify()` doesn't work with the 'dummy' video driver. This means you need to split the test into two parts and add a skip decorator to the top. that looks something like this:

```
    @unittest.skipIf(
        os.environ.get("SDL_VIDEODRIVER") == "dummy",
        'requires the SDL_VIDEODRIVER to be a non "dummy" value',
    )
    def test_get_active_iconify(self):
```

3. I'm not sure what the issue is on Ubuntu but if the original test code was working for you and the test above doesn't then it sounds like it might be related to calling `display.update()` and `event.pump()` a few times so I would be tempted to just a a for loop that calls those functions a few times after calling iconify and seeing if that resolves the difference. I'll have a look on raspberry pi later myself.

**Comment by tsadama** [comment]
Hi I made the changes requested above, but I am still getting the issue where my CI builds are terminating because of a lack of output for 10 minutes. The previous 3 commits were to make sure everything is updated with my fork/branch, but I don't know where the issue stems from.

**Comment by MyreMylar** [comment]
From what I can tell Travis is failing on the SDL1 build on line 74. It looks to me it like there is an arguable bug in SDL 1 there. It appears that:

` return PyBool_FromLong((SDL_GetAppState() & SDL_APPACTIVE) != 0);`

is returning True before a window has even been created there.  

I think we could make the case for changing SDL 1's get_active to something like:

```
#else  /* IS_SDLv1 */
static PyObject *
pg_get_active(PyObject *self, PyObject *args)
{
    if (!pgDisplaySurfaceObject)
         return PyBool_FromLong(0);
    return PyBool_FromLong((SDL_GetAppState() & SDL_APPACTIVE) != 0);
}
#endif /* IS_SDLv1 */
```

to fit better with how the docs describe the function working only after set_mode has been called.

Both of these new tests passed on travis on SDL2 so if there is something else up with the Travis builds it is almost certainly not your fault. It looks like a whole bunch of new pull requests were merged so perhaps one of them has sewed chaos. 

Appveyor hasn't caught up with this PR yet either, but if it passes on there it is probably just Travis having a bad day which happens occasionally. I would try to fix the SDL1 problem either by changing the display.c code (better) or skipping that part of the test on SDL1 (OK). You shousl be able to test on SDL1 locally by changing the build config andthen doing a clean and rebuild, something like:

```
python -m buildconfig -sdl1
python setup.py clean --all
python setup.py install
pip install . -U 
```

Or however you do that process on your platform/build environment. You can switch back to sdl2 by doing the same thing but with sdl2 in place of sdl1 on the first line.

File: test/display_test.py [hunk 1]
```diff
@@ -65,20 +65,53 @@ def todo_test_flip(self):
 
         self.fail()
 
-    def todo_test_get_active(self):
+    def test_get_active(self):
+        """Test the get_active function"""
+
+        #Initially, the display is not active
+        pygame.quit()
```
Comment (MyreMylar) [inline test/display_test.py:72]: don't need this pygame.quit() here. The display tests only init the display module before each test with the setUp() function:

https://github.com/pygame/pygame/blob/4940b9c664f8b063e8f634857c9f79ec5eb736d7/test/display_test.py#L18-L19

File: test/display_test.py [hunk 2]
```diff
@@ -65,20 +65,53 @@ def todo_test_flip(self):
...
+        pygame.display.quit()
+        self.assertEqual(pygame.display.get_active(), False)
+
+        #get_active defaults to true after a set_mode
+        pygame.init()
```
Comment (MyreMylar) [inline test/display_test.py:77]: likewise should change this to `pygame.display.init()` since we are just testing the display module.

File: test/display_test.py [hunk 3]
```diff
@@ -65,20 +65,53 @@ def todo_test_flip(self):
...
+
+
+    @unittest.skipIf(
+        os.environ.get("SDL_VIDEODRIVER") == "dummy",
+        'requires the SDL_VIDEODRIVER to be a non dummy value'
+    )
+    def test_get_active_iconify(self):
+        """Test the get_active function after an iconify"""
+
+        pygame.init()
```
Comment (MyreMylar) [inline test/display_test.py:95]: should delete this as `setUp()` will already have called `pygame.display.init()` for us.

File: test/display_test.py [hunk 4]
```diff
@@ -416,20 +449,42 @@ def todo_test_set_palette(self):
 
         self.fail()
 
-    def todo_test_toggle_fullscreen(self):
+    def test_toggle_fullscreen(self):
```
Comment (MyreMylar) [inline test/display_test.py:452]: toggle_fullscreen() doesn't work with the dummy video driver. We should skip this whole test on CI.

According to the docs it also only works on X11 (linux) - this is clearly untrue looking at the current code which indicates support varies widely on a per platform basis. Probably we should also raise an issue to try and update the docs with more accurate info on what platforms toggle_fullscreen() now works on.

Two approaches here, you could either parse through the C code to try and work out which platforms currently support toggle_fullscreen(). This is the relevant part of the code:

https://github.com/pygame/pygame/blob/7d3900d5b04dc8102f218cc60f3fbf5e06dd8fa1/src_c/display.c#L2515-L2559

And then add checking for those platforms as part of the test (looks like it should work on windows, mac (cocoa) and linux (x11 & wayland drivers) at least but fail on android which is always full screen). Or, probably better, you could just make the test fail if the function raises an unexpected exception (as I mention below) and then build up a skip list of list of platforms that the test would fail on.

To figure out platforms for this you can probably mostly rely on the SDL_VIDEODRIVER environment variable again which also has 'android' in addition to the 'dummy' on used on CI. See this list in the get_driver test:

https://github.com/pygame/pygame/blob/48e94ef6106f3c1b79f95e0012f8f7ac2a964066/test/display_test.py#L106-L137

Comment (MyreMylar) [reply test/display_test.py:452]: TLDR: skip the test on dummy SDL_VIDEODRIVER and android at least.

**Comment by MyreMylar** [comment]
I think probably the toggle fullscreen problems are why this is causing the CI to lock up.

**Comment by MyreMylar** [comment]
I expect there may still be some edge cases where the `toggle_fullscreen()` test fails but given it only runs locally and most of us don't personally have the many potential configurations & platforms that pygame could run under I think it might be easier to see if we get any reports of failures for this test and then figure out how to skip the test in that particular situation.

**Comment by MyreMylar** [comment]
Checked the current commit on windows under SDL1 & SDL2 and all the tests pass.

I'll submit the change to SDL2's version of get_active() as a separate PR, unless you want to roll it into here?

**Comment by tsadama** [comment]
I don't mind if you submit the SDL2 as a separate PR! Thank you for all the help @MyreMylar !

**Comment by MyreMylar** [comment]
Added issue about updating the `toggle_fullscreen()` docs here: https://github.com/pygame/pygame/issues/1990

**Comment by MyreMylar** [comment]
Just holding off merging this until the windows fix can get merged at the same time.
